File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/utils/PressAction.java
package org.firstinspires.ftc.teamcode.utils;

import java.util.function.Supplier;

/**
 * A delegate for single press buttons.
 * <p>
 * When the button is pressed, the action will run once.
 */
public class PressAction {
    private boolean lastState = false;
    private final Supplier<Boolean> getState;
    private final Runnable action;
    /**
     * Creates a new PressAction delegate.
     * @param getState a supplier for the button state.
     * @param action the action to run when the button is pressed.
     */
    public PressAction(Supplier<Boolean> getState, Runnable action) {
        this.getState = getState;
        this.action = action;
    }
    /**
     * Runs the action if the button is pressed.
     */
    public void run() {
        boolean state = getState.get();
        if (state && !lastState) action.run();
        lastState = state;
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/utils/PressAction.kt
package org.firstinspires.ftc.utils

class PressAction(
    /**
     * The button to check.
     */
    private val button: () -> Boolean,
    /**
     * The action to run when the button is pressed.
     */
    private val action: () -> Unit
) {
    // The last state of the button
    private var lastState = false
    /**
     * Runs the action if the button is pressed.
     */
    fun run() {
        val state = button()
        if (state && !lastState) {
            action()
        }
        lastState = state
    }
}